### DISCUSSION TITLE
The Challenges of Sub-Elements ( [el , [tr , [td , etc.)

### DISCUSSION THREAD

**Discussion by tajmone** [pml-lang/pml-companion#37] [Editors Support]
There are some nodes that pose various challenges in the implementation of a PML syntax for editors, i.e. those nodes which are designed to be sub-nodes of another node — e.g. list element `[el`, which should occur inside a `[list`; and similarly all the `[table` sub-nodes (`[tr`, `[tc`, etc.).

Instinctively, it's tempting to implement them in such a way that `[el` nodes will only be recognized as valid elements when occurring inside a `[list`, and to scope as `invalid` any `[el` nodes occurring in the root context.

But the question begs: are lists nestable? What type of block nodes can a `[list` contain?

The documentation doesn't delve much into these details and, without a BNF grammar of the PML syntax, devs can only attempt trial and error cases to shed light on the issue. But whether lists are (or will eventually be) nestable can make a huge difference on how `[list` and `[el` should be implemented in an editor syntax.

Also, while it might at first seem reasonable to scope as `invalid` any element occurring out of its natural context, we need to remember that the use of `[insert` could result in PML snippet-files containing only lists elements (`[el`) or table inner element (`[tr`, `[tc`, etc.), which would then be scoped as `invalid` (i.g. `invalid.out-of-context`, or something like that). Is that what we really want?

The availability of the `[insert` node is one of the main reasons that make me reluctant in enforcing strict nodes-context validation in Sublime PML, because I can easily envision snippet files containing partial sub-elements of all kinds.

The lack of a BNF, and the fact that many elements could change in the future, are also factors which are holding me back from implementing some syntax elements which can be challenging — I want to avoid at all costs that future changes could break the ST syntax, because all these RegEx-based Stack states can too easily crumble to pieces if any element radically changes behaviour.

But a full knowledge of which specific node (block and inline alike) are permitted inside each node (especially block nodes) is critical for the syntax implementation. So far, I've been relying on trial and error testing using `pmlc`, but the possible edge cases to cover are really huge in number.

E.g. if we want to ensure that an `[el` node is marked as `invalid` when it occurs outside a list, then we need to capture it as invalid in the root context, and as valid within a `[list` context — which in turn means that the `[list` context needs to support (include) all the nodes that are eligible as it sub-elements, and mark as `invalid` those which aren't.

As already mentioned elsewhere, there's a major syntax rework awaiting here, because once all the nodes are implemented I'll need to revise the whole syntax to ensure that valid sub-nodes are matched via their explicit inclusion in each parent node, so that all other nodes can be matched as `invalid`.

Needless to say, it's always a bit scary to face a similar syntax rewrite, because things could break badly at any stage. Now, Sublime Text offers a powerful system to test syntax scopes, so the included PML test suite is a safety anchor to prevent total breakage. But many editors that adopt a similar scoping approach (_a la_ TextMate, to be clear) don't provide similar tools — e.g. VSCode offers syntax debugging, but it's not the same thing, and doesn't allow for fine grain testing with a simple keystroke to auto-detect any breaking changes. So I'm no quite sure how the same syntax granularity could be achieved in other editors without loosing sight of the great picture and incurring in bugs.

**Comment by pml-lang** [comment]
> are lists nestable?

Yes. And there is no limit as to the depth of nesting.

> What type of block nodes can a [list contain?

There are currently no rules applied by the parser. If elements are nested incorrectly then the behavior is undefined.
Example: `[el` nodes without a `[list` as parent
The converter should emit an error, but doesn't (yet).  It will create HTML `li` elements without embedding them in `ul`.
Hence the outcome depends on the browser. In Firefox, for example, the elements are displayed, but not indented.

> scope as invalid any element occurring out of its natural context

That would be really great, but I think it's only doable in the context of using [LSP](https://microsoft.github.io/language-server-protocol/).

> the use of [insert could result in PML snippet-files containing only lists elements ([el) or table inner element ([tr, [tc, etc.), which would then be scoped as invalid

Yes, that's the main challenge.

> The availability of the [insert node is one of the main reasons that make me reluctant in enforcing strict nodes-context validation in Sublime PML

Exactly! That's why it would be very difficult (if not impossible) to reliably add this kind of validation in an editor without using LSP.

> many elements could change in the future, are also factors which are holding me back from implementing some syntax elements which can be challenging

True. Once these rules will be defined and implemented in the new parser, they risk to change often (especially at the beginning, and when new nodes are added). Each change would require the editor plugin to be updated. Maintenance nightmare!
One could of course define the rules in a standard format (e.g. XML Schema). But then the editor plugin must be able to read the schema, and apply the rules.
'Tree structure validation' must surely be implemented in a future version of the converter, but just now no decision has been taken yet as to  _how_ to do it and which format to use.

> it's always a bit scary to face a similar syntax rewrite, because things could break badly at any stage.

I agree, and therefore IMHO I think that tree structure validation should currently not be implemented in any editor plugin.

**Reply by tajmone** [reply]
> > scope as invalid any element occurring out of its natural context
>
> That would be really great, but I think it's only doable in the context of using LSP.

No, it's doable is TexMate like syntaxes that support pushing/popping contexts on the Stack: you'd only need to match `[el` as invalid when occurring in the root context or in any context not nested into `[list`; the `[list` context would capture `[el` as a valid node. It's not different than handling nested comments, which are just recursive contexts that only expect a closing comment tag or another comment start.

The problem here is just that in standalone snippets intended for `[insert`ion you could fin loose `[el` elements.

The only solution that comes to mind is differentiating between sources that have a `[doc` node and those which don't; i.e. the latter being certainly snippets intended for `[insert`ion into other docs. But this might not cover all cases either, for in theory a document (i.e. a source with `[doc`) could be omitting the `[list` element because it's being `[insert`ed by a snippet (although this would be a rare case).

In any case, handling the double standard of source with and without `[doc` adds come complexity to the syntax, which might make it a bit harder to track (maybe not so much in ST, but more so in editors that use simpler JSON definitions, like VSCode).

## LSP Might Not Handle These Either...

> > The availability of the `[insert` node is one of the main reasons that make me reluctant in enforcing strict nodes-context validation in Sublime PML
>
> Exactly! That's why it would be very difficult (if not impossible) to reliably add this kind of validation in an editor without using LSP.

I'm not sure that LSP could entirely solve this problem either. It largely depends on whether the editor supports "projects", and its ability to decide whether different PML sources belong to a same project or not.

Also, the fact that URIs (and file paths) can contain parameters, is going to make it hard even for an LSP syntax to track files `[insert`ions, especially since insertion paths are relative to the converted doc location. And, even _if_ it could possible to resolve all parameters expansions correctly, and determine all insertion paths and which documents are inserted where, this would only work for completed projects, and not cover any WIP — if gracefully handling incomplete syntax constructs is one of LSP's main challenges, this is yet another challenge that add layers of complications on how to handle incomplete assets and update their references as they are added/completed.

Furthermore, all the above consideration are assuming that parameters are defined within a PML source in the project, and not via a script (or worst: manually via CLI invocation of PML), and that there are no conditional factors that might allow for different document resolution based on different constant values — which, I believe, is something that will ultimately be added to PML, since it's a powerful feature that allows to control aspects of a publication depending on the format, the publisher, the document edition (free sample vs full product, etc.).

That is to say that we need to be cautious when considering how far any syntax definition (LSP or otherwise) should go in trying to validate the source syntax and/or try to track parameters expansions, URIs and documents `[insert`ions. The fact that it might be doable doesn't necessarily mean it's desirable, or that it will still be doable when new features are added to PML.

The problem with editor syntaxes is that often they are created quickly in the desire to cover the syntax as soon as possible, but often they end up breaking up badly when new changes to the language are introduced, or just because they were built without a proper test suite that could track edge cases and scopes disruption. I've seen this happen too many times, which is why I've decided that for PML I would take a cautious and well planned approach, never adding new elements until the previous ones are well covered by tests.

A small change in design (like the tag boundaries discussed in #39) could easily break a whole syntax — and, in any case, the bigger the test-suite, the more work to re-adapt all the (now failing) tests to the new changes. Testing syntax scopes is all about ensuring that the syntax is behaving as expected, so whenever a scope change is introduced, you can't just change all the tests scopes just to make the test pass; you need to ensure that the new scopes don't violate the expected behaviour.

## Should Editor Syntaxes Validate?

So, as far as handling elements like `[el`, `[tr`, etc. goes, the problem is deciding whether attempting to track invalid nodes usage is worth the effort, considering how advanced `[insert` and parameters uses is going to limit similar attempts. My original intention was not to expect the editor syntax to cover a validation role at all — i.e. allow any attribute in any node, as long as it occurs in the rightfully allocated area after a node, but without worrying about whether the node actually supports it.

Although I now it will be possible to mark as invalid any unsupported attribute in a node (once all nodes and attributes are implemented), I'm still not fully convinced that doing so would be a good idea.

Usually, syntaxes don't care with such details (except for the most obvious invalid cases) but rely instead on the compiler/builder tool to validate source files — just like builds are able to map error messages to their source files. I guess that eventually PMLC will offer a dry-run syntax validation option, so might focus on that instead of syntax predictions.

Good LSP Lang Servers do that anyway, especially if the language supports incremental compiling. Compilers are a more reliable source of information when it comes to validating source files. Furthermore, they allow users to work with older tools that will validate a source file according to a given version, whereas a syntax definition can only cover the semantics of a specific language release.

For Sublime PML, instead of validating attributes according to node contexts, we could just accept any well-formed attribute in any node that supports attributes, and think of creating a Python script that checks whether the source file contains any node with unsupported attributes — although I don't see how such a script could be better than just dry-running PMLC on the source files.

**Reply by pml-lang** [reply]
> I'm not sure that LSP could entirely solve this problem either.

As far as I understand (with my 0 LSP experience) a possibly viable solution covering all the challenges well mentioned in your comment is this:

The language server:
- uses the PML parser API (i.e. a customized version for LSP) to handle PML features like `const`, `insert`, conditional parsing, etc. 
- uses the LSP protocol (JSON-RPC) to communicate with text editors

**Reply by tajmone** [reply]
But even with that coverage, the LSP Lang Server wouldn't be able to resolve parameters which are defined via the command line (when the feature is implemented). E.g. a parameter constant might be defined to use different suffixes in images filenames (or their paths folders) for different editions of a same document (e.g. large vs small images) or different extension (e.g. PNG vs SVG, etc.).

In a complex toolchain that can build different versions of a same document, for different target audiences (e.g. an author who published the same article book on different websites), parts of the document will conditionally depend on some parameter value being defined at invocation time (e.g. different ISBN number, publisher info, etc.).

For a good example of a software that natively supports automatically publishing a book in multiple editions, targeting not only different formats but also different online sellers (e.g. Lulu, Kobo, Apple Store, etc.) see Jutoh:

http://www.jutoh.com

This is a great tool that allows one-click publishing to convert a book in different formats and editions, and it covers all the required metadata expected by book publishers and online stores. It's also capable of building different versions of the same eBook targeting different eBook readers, thanks to an internal database containing their screen resolution (e.g. adjust cover and images sizes accordingly) and known quirks in how they support ePub features (i.e. adapting the generated XML to meet those quirks).

If you install the trial version you can read through the User Guide, which illustrates how these features works.

The Jutoh Plus edition (that's the one I bought) also supports further control via advanced scripting, to customize the build toolchain:

http://www.jutoh.com/jutohplus.html

Jutoh is a good reference for how publisher oriented software should be developed. Unfortunately Jutoh project files are stored in a proprietary binary format (SQLite databases) which hinders version controlled collaborative editing, which is probably why it hasn't become a widely adopted tool in the industry, despite all its powerful features.

**Reply by pml-lang** [reply]
> resolve parameters which are defined via the command line

In that case the parameters would need to be passed to the language server as well.

> parts of the document will conditionally depend on some parameter value

Yes, and with the new text template engine I intend to integrate in PML, PML code can be intermixed with template code written in Javascript, Python, or any other language supported by [GraalVM](https://www.graalvm.org/reference-manual/embed-languages/#access-guest-languages-directly-from-java). This can be used for example to implement conditional parsing, or to generate PML code programmatically. However, I think that template code sections can easily be ignored (like comments), because this code will be embedded between a `{{...}}` pair or similar. That is, unless we want to switch syntax highlighting to the language used in the template.

Thanks a lot for the link to Jutoh.

**Reply by tajmone** [reply]
> That is, unless we want to switch syntax highlighting to the language used in the template.

This should be easy to do in most editors and syntax highlighters. Sublime PML already hands over the contents of an `[html` node to ST's native HTML highlighter, so the raw HTML block is highlighted too.

Since the template engine uses popular syntaxes like JS and Python, there's no reason to suspect that an editor won't already provide native highlighting for those. As for less-known syntaxes, or syntaxes which require supporting PML template references, you can always include in the PML syntax plug-in an _ad hoc_ syntax to parse the `{{...}}` template contents according to custom rules.

**Reply by pml-lang** [reply]
> Sublime PML already hands over the contents of an [html node to ST's native HTML highlighter, so the raw HTML block is highlighted too.

Great!

**Reply by celtic-coder** [reply]
> > are lists nestable?
> 
> Yes. And there is no limit as to the depth of nesting.
> 
> > What type of block nodes can a [list contain?
> 
> There are currently no rules applied by the parser. If elements are nested incorrectly then the behavior is undefined.
> Example: `[el` nodes without a `[list` as parent
> The converter should emit an error, but doesn't (yet). It will create HTML `li` elements without embedding them in `ul`.
> Hence the outcome depends on the browser. In Firefox, for example, the elements are displayed, but not indented.

I tried the following experiment:

![sub-list-testing](https://user-images.githubusercontent.com/11787866/124291990-0d7a0f80-db4d-11eb-8288-07074f0366a5.png)

However, I am getting the following error:

![only-el-elements](https://user-images.githubusercontent.com/11787866/124292397-6e094c80-db4d-11eb-89ef-e97051c22f14.png)

Further, if I nest an `[el` element within another `[el` element, I just get a list of items all at the same level. Is the syntax that I am using for nestable lists incorrect, perhaps?

**Reply by celtic-coder** [reply]
Hmm... I have resolved the issue. It turns out that the `[list` element must be enclosed in a parent `[el` element. It's always worth experimenting in order to find a solution!

**Reply by tajmone** [reply]
> I have resolved the issue. It turns out that the `[list` element must be enclosed in a parent `[el` element.

Do you mean like this?

```
[list
    [el Element]
    [el
        [list
            [el Sub-Element]
        ]
    ]
]
```

That would be the way HTML handles lists — which has always been a bit confusing to beginners, but correct when you come to think about.

> It's always worth experimenting in order to find a solution!

Indeed, when I have to add a new syntax element to Sublime PML I always end up having to do lot's of tests to try and work out how nodes behave in edge-cases, in combinations with other nodes, etc. You often discover unexpected things, which are not mentioned in the documentation.

Also, when it comes to creating a syntax highlighter things are more complex, because you have to deal with subtle aspects of nodes and attributes (e.g. see tajmone/Sublime-PML#26 on the `[code`).

But I still think it's worth documenting edge case behaviours, at least in asides or footnotes. Also, it might be useful to add such examples somewhere in the [PML Playground] repository, mainly to ensure that future updates to the language don't change these edge cases.

If you look a the syntax test of Sublime PML, you'll find many nodes' edge cases and quirks being covered. Without these tests, it would be hard to detect if changes to the language are breaking how the package parses nodes.



[PML Playground]: https://github.com/tajmone/pml-playground

**Reply by tajmone** [reply]
@pml-lang, is there an ETA for when a PML BNF document will be available?

It doesn't have to be parser-proof, just a representation of how the syntax is structured. Having a BNF of some sort would be very useful for developers working on highlighters, converters, etc.

**Reply by pml-lang** [reply]
> is there an ETA for when a PML BNF document will be available?

Please have a look at the syntax of the upcoming PML version 2.0 [here](https://www.pml-lang.dev/docs/syntax/ebnf.html).

Note that this is a _draft_. It is not official yet (you can't access it by a link from PML's website). Moreover, the rules of lenient parsing are not yet described.

I will review this EBNF in the coming days, and add information about lenient parsing.

I hope this helps.
